File: java/src/main/java/vsdk/toolkit/io/wrapper/FileUncompressWrapper.java
package vsdk.toolkit.io.wrapper;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import vsdk.toolkit.common.Error;

public class FileUncompressWrapper {
    public static InputStream openInputStreamCompressWrapper(String fileName, int[] isPipe) {
        if (isPipe != null && isPipe.length > 0) {
            isPipe[0] = 0;
        }
        if (isInvalidFileName(fileName)) {
            return null;
        }

        int commandLength = fileName.length() + 20;
        StringBuilder command = new StringBuilder(commandLength);
        boolean pipeFlag = buildPipeCommand(fileName, StreamOpenMode.READ, command, commandLength);

        InputStream stream = null;
        if (pipeFlag) {
            stream = openPipeInputStream(command.toString());
        }
        else {
            File file = new File(fileName);
            if (file.exists() && file.canRead() && file.isFile()) {
                try {
                    stream = new FileInputStream(fileName);
                }
                catch (Exception ignored) {
                    stream = null;
                }
            }
        }

        if (stream == null) {
            Error.error(null, "Can't open file '%s' for %s", fileName, modeToLogAction(StreamOpenMode.READ));
            if (isPipe != null && isPipe.length > 0) {
                isPipe[0] = 0;
            }
            return null;
        }

        if (isPipe != null && isPipe.length > 0) {
            isPipe[0] = pipeFlag ? 1 : 0;
        }
        return stream;
    }

    public static OutputStream openOutputStreamCompressWrapper(String fileName, int[] isPipe) {
        if (isPipe != null && isPipe.length > 0) {
            isPipe[0] = 0;
        }
        if (isInvalidFileName(fileName)) {
            return null;
        }

        int commandLength = fileName.length() + 20;
        StringBuilder command = new StringBuilder(commandLength);
        boolean pipeFlag = buildPipeCommand(fileName, StreamOpenMode.WRITE, command, commandLength);

        OutputStream stream = null;
        if (pipeFlag) {
            stream = openPipeOutputStream(command.toString());
        }
        else {
            File file = new File(fileName);
            if (!file.isDirectory()) {
                try {
                    stream = new FileOutputStream(fileName);
                }
                catch (Exception ignored) {
                    stream = null;
                }
            }
        }

        if (stream == null) {
            Error.error(null, "Can't open file '%s' for %s", fileName, modeToLogAction(StreamOpenMode.WRITE));
            if (isPipe != null && isPipe.length > 0) {
                isPipe[0] = 0;
            }
            return null;
        }

        if (isPipe != null && isPipe.length > 0) {
            isPipe[0] = pipeFlag ? 1 : 0;
        }
        return stream;
    }

    public static void closeInputStream(InputStream stream) {
        if (stream == null) {
            return;
        }
        try {
            stream.close();
        }
        catch (Exception ignored) {
        }
    }

    public static void closeOutputStream(OutputStream stream) {
        if (stream == null) {
            return;
        }
        try {
            stream.close();
        }
        catch (Exception ignored) {
        }
    }

    private static String modeToLogAction(StreamOpenMode mode) {
        return mode == StreamOpenMode.READ ? "reading" : "writing";
    }

    private static boolean isInvalidFileName(String fileName) {
        if (fileName == null || fileName.isEmpty() || fileName.endsWith("/")) {
            return true;
        }
        return false;
    }

    private static boolean buildPipeCommand(String fileName, StreamOpenMode openMode, StringBuilder command, int commandLength) {
        if (fileName == null || command == null || commandLength <= 0) {
            return false;
        }

        command.setLength(0);

        if (fileName.charAt(0) == '|') {
            command.append(fileName.substring(1));
            return true;
        }

        int dot = fileName.lastIndexOf('.');
        String ext = dot >= 0 ? fileName.substring(dot) : null;
        if (".gz".equals(ext)) {
            if (openMode == StreamOpenMode.READ) {
                command.append("gunzip < ").append(fileName);
            }
            else {
                command.append("gzip > ").append(fileName);
            }
        }
        else if (".Z".equals(ext)) {
            if (openMode == StreamOpenMode.READ) {
                command.append("uncompress < ").append(fileName);
            }
            else {
                command.append("compress > ").append(fileName);
            }
        }
        else if (".bz".equals(ext)) {
            if (openMode == StreamOpenMode.READ) {
                command.append("bunzip < ").append(fileName);
            }
            else {
                command.append("bzip > ").append(fileName);
            }
        }
        else if (".bz2".equals(ext)) {
            if (openMode == StreamOpenMode.READ) {
                command.append("bunzip2 < ").append(fileName);
            }
            else {
                command.append("bzip2 > ").append(fileName);
            }
        }
        else {
            return false;
        }
        return true;
    }

    private static InputStream openPipeInputStream(String command) {
        PipeInputStream pipeStream = new PipeInputStream(command);
        if (!pipeStream.isOpen()) {
            return null;
        }
        return pipeStream;
    }

    private static OutputStream openPipeOutputStream(String command) {
        PipeOutputStream pipeStream = new PipeOutputStream(command);
        if (!pipeStream.isOpen()) {
            return null;
        }
        return pipeStream;
    }
}

File: turboc/src/io/wrapper/FileUncompressWrapper.cpp
#include <string.h>

#include "java/io/FileInputStream.h"
#include "java/io/FileOutputStream.h"
#include "java/util/Formatter.h"
#include "common/logging/Logger.h"
#include "io/wrapper/FileUncompressWrapper.h"
#include "io/wrapper/PipeInputStream.h"
#include "io/wrapper/PipeOutputStream.h"
#include "io/wrapper/StreamOpenMode.h"

const char *
FileUncompressWrapper::modeToLogAction(StreamOpenMode mode) {
    return mode == READ ? "reading" : "writing";
}

bool
FileUncompressWrapper::isInvalidFileName(const char *fileName) {
    if ( fileName == NULL || fileName[0] == '\0' || fileName[strlen(fileName) - 1] == '/' ) {
        return true;
    }
    return false;
}

bool
FileUncompressWrapper::buildPipeCommand(const char *fileName, StreamOpenMode openMode, char *command, int commandLength) {
    if ( fileName == NULL || command == NULL || commandLength <= 0 ) {
        return false;
    }

    if ( fileName[0] == '|' ) {
        Formatter::format(command, commandLength, "%s", &fileName[1]);
        return true;
    }

    const char *ext = strrchr(fileName, '.');
    if ( ext && strcmp(ext, ".gz") == 0 ) {
        if ( openMode == READ ) {
            Formatter::format(command, commandLength, "gunzip < %s", fileName);
        } else {
            Formatter::format(command, commandLength, "gzip > %s", fileName);
        }
    } else if ( ext && strcmp(ext, ".Z") == 0 ) {
        if ( openMode == READ ) {
            Formatter::format(command, commandLength, "uncompress < %s", fileName);
        } else {
            Formatter::format(command, commandLength, "compress > %s", fileName);
        }
    } else if ( ext && strcmp(ext, ".bz") == 0 ) {
        if ( openMode == READ ) {
            Formatter::format(command, commandLength, "bunzip < %s", fileName);
        } else {
            Formatter::format(command, commandLength, "bzip > %s", fileName);
        }
    } else if ( ext && strcmp(ext, ".bz2") == 0 ) {
        if ( openMode == READ ) {
            Formatter::format(command, commandLength, "bunzip2 < %s", fileName);
        } else {
            Formatter::format(command, commandLength, "bzip2 > %s", fileName);
        }
    } else {
        return false;
    }
    return true;
}

InputStream *
FileUncompressWrapper::openPipeInputStream(const char *command) {
    PipeInputStream *pipeStream = new PipeInputStream(command);
    if ( !pipeStream->isOpen() ) {
        delete pipeStream;
        return NULL;
    }
    return pipeStream;
}

OutputStream *
FileUncompressWrapper::openPipeOutputStream(const char *command) {
    PipeOutputStream *pipeStream = new PipeOutputStream(command);
    if ( !pipeStream->isOpen() ) {
        delete pipeStream;
        return NULL;
    }
    return pipeStream;
}

InputStream *
FileUncompressWrapper::openInputStreamCompressWrapper(const char *fileName, int *isPipe) {
    if ( isPipe != NULL ) {
        *isPipe = 0;
    }
    if ( isInvalidFileName(fileName) ) {
        return NULL;
    }

    const int commandLength = ((int)(strlen(fileName))) + 20;
    char *command = new char[commandLength];
    const bool pipeFlag = buildPipeCommand(fileName, READ, command, commandLength);

    InputStream *stream = NULL;
    if ( pipeFlag ) {
        stream = openPipeInputStream(command);
    } else {
        File file(fileName);
        if ( file.exists() && file.canRead() && file.isFile() ) {
            stream = new FileInputStream(fileName);
        }
    }
    delete[] command;

    if ( stream == NULL ) {
        Logger::error(NULL, "Can't open file '%s' for %s", fileName, modeToLogAction(READ));
        if ( isPipe != NULL ) {
            *isPipe = 0;
        }
        return NULL;
    }

    if ( isPipe != NULL ) {
        *isPipe = pipeFlag ? 1 : 0;
    }
    return stream;
}

OutputStream *
FileUncompressWrapper::openOutputStreamCompressWrapper(const char *fileName, int *isPipe) {
    if ( isPipe != NULL ) {
        *isPipe = 0;
    }
    if ( isInvalidFileName(fileName) ) {
        return NULL;
    }

    const int commandLength = ((int)(strlen(fileName))) + 20;
    char *command = new char[commandLength];
    const bool pipeFlag = buildPipeCommand(fileName, WRITE, command, commandLength);

    OutputStream *stream = NULL;
    if ( pipeFlag ) {
        stream = openPipeOutputStream(command);
    } else {
        File file(fileName);
        if ( file.canWrite() && !file.isDirectory() ) {
            stream = new FileOutputStream(fileName);
        }
    }
    delete[] command;

    if ( stream == NULL ) {
        Logger::error(NULL, "Can't open file '%s' for %s", fileName, modeToLogAction(WRITE));
        if ( isPipe != NULL ) {
            *isPipe = 0;
        }
        return NULL;
    }

    if ( isPipe != NULL ) {
        *isPipe = pipeFlag ? 1 : 0;
    }
    return stream;
}

void
FileUncompressWrapper::closeInputStream(InputStream *stream) {
    if ( stream == NULL ) {
        return;
    }
    stream->close();
    delete stream;
}

void
FileUncompressWrapper::closeOutputStream(OutputStream *stream) {
    if ( stream == NULL ) {
        return;
    }
    stream->close();
    delete stream;
}
